crates/re_viewer/src/web_tools.rs

```diff
@@ -37,141 +80,278 @@ pub fn set_url_parameter_and_refresh(key: &str, value: &str) -> Result<(), wasm_
+
+        // the user navigated back to the history entry where the viewer was initially opened
+        // in that case they likely expect to land back at the welcome screen:
+        command_sender.send_system(SystemCommand::CloseAllRecordings);
+        command_sender.send_system(SystemCommand::ActivateApp(StoreHub::welcome_screen_app_id()));
+
+        set_stored_history_entry(new_state);
+        egui_ctx.request_repaint();
+
+        return Ok(());
+    }
+
+    let Some(entry) = new_state else {
+        unreachable!();
+    };
+
+    let follow_if_http = false;
+    for url in &entry.urls {
+        // we continue in case of errors because some receivers may be valid
+        let Some(receiver) =
+            url_to_receiver(egui_ctx.clone(), follow_if_http, url.clone()).ok_or_log_error()
+        else {
+            continue;
+        };
+
+        command_sender.send_system(SystemCommand::AddReceiver(receiver));
+
+        re_log::debug!("popstate: add receiver {url:?}");
+    }
+
+    set_stored_history_entry(Some(entry));
+    egui_ctx.request_repaint();
+
+    Ok(())
 }
 
 pub fn go_back() -> Option<()> {
-    let history = web_sys::window()?
-        .history()
-        .map_err(|err| format!("Failed to get History API: {}", string_from_js_value(err)))
-        .ok_or_log_error()?;
-    history
-        .back()
-        .map_err(|err| format!("Failed to go back: {}", string_from_js_value(err)))
-        .ok_or_log_error()
+    let history = history().ok_or_log_js_error()?;
+    history.back().ok_or_log_js_error()
 }
 
 pub fn go_forward() -> Option<()> {
-    let history = web_sys::window()?
-        .history()
-        .map_err(|err| format!("Failed to get History API: {}", string_from_js_value(err)))
-        .ok_or_log_error()?;
-    history
-        .forward()
-        .map_err(|err| format!("Failed to go forward: {}", string_from_js_value(err)))
-        .ok_or_log_error()
+    let history = history().ok_or_log_js_error()?;
+    history.forward().ok_or_log_js_error()
 }
 
-/// The current percent-encoded URL suffix, e.g. "?foo=bar#baz".
-pub fn current_url_suffix() -> Option<String> {
-    let location = web_sys::window()?.location();
-    let search = location.search().unwrap_or_default();
-    let hash = location.hash().unwrap_or_default();
-    Some(format!("{search}{hash}"))
+/// A history entry is actually stored in two places:
+/// - State object
+/// - URL
+///
+/// Ideally we wouldn't have to, but we want two things:
+/// - Listen to `popstate` events and handle navigations client-side,
+///   so that the forward/back buttons can be used to navigate between
+///   examples and the welcome screen.
+/// - Add a `?url` query param to the address bar when navigating to
+///   an example, so that examples can be shared directly by just
+///   copying the link.
+#[derive(Clone, Default, Debug, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
+pub struct HistoryEntry {
```

Comment:
This is a very useful comment; unfortunately it comes quite late in the file (after the undocumented `STORED_HISTORY_ENTRY`). I think this file would be easier to parse if `HistoryEntry` came earlier up.

It might also be a good idea to just move all the history stuff to its own file (a sub-module to `web_tools`) and have the comment as a module-level docstring.